clarity/src/vm/callables.rs

```diff
@@ -168,29 +170,39 @@ impl DefinedFunction {
         for arg in arg_iterator.drain(..) {
             let ((name, type_sig), value) = arg;
 
-            match (type_sig, value) {
+            // Arguments containing principal literals can be implicitly cast to traits
+            // to match parameter types.
+            // e.g. `(some .foo)` to `(optional <trait>`)
+            // and traits can be implicitly cast to sub-traits
+            // e.g. `<foo-and-bar>` to `<foo>`
+            let cast_value = implicit_cast(type_sig, value)?;
+
+            match (&type_sig, &cast_value) {
                 (
-                    TypeSignature::TraitReferenceType(trait_identifier),
-                    Value::Principal(PrincipalData::Contract(callee_contract_id)),
+                    _,
+                    Value::Trait(TraitData {
+                        contract_identifier,
+                        trait_identifier,
+                    }),
                 ) => {
-                    // Argument is a trait reference, probably leading to a dynamic contract call
+                    // Argument is a trait reference, probably leading to a dynamic contract call.
                     // We keep a reference of the mapping (var-name: (callee_contract_id, trait_id)) in the context.
-                    // The code fetching and checking the trait is implemented in the contract_call eval function.
+                    // The trait compatibility has been checked by the type-checker.
                     context.callable_contracts.insert(
                         name.clone(),
-                        (callee_contract_id.clone(), trait_identifier.clone()),
+                        TraitData {
+                            contract_identifier: contract_identifier.clone(),
+                            trait_identifier: trait_identifier.clone(),
+                        },
                     );
                 }
```

Comment:
Here, I'm inclined to say that a `Callable` value should only remain callable if the argument was a trait (and the trait should do a upcast).

For example:

```clar
;; <trait-a> is a super-type of <trait-b>
;;  i.e., if Contract-B impls <trait-b> and Contract-A impls <trait-a>,
;;          then Contract-B impls <trait-a> as well, but Contract-A does not impl <trait-b>

(define-private (foo-0 (impl-contract <trait-b>))
   (foo-1 impl-contract))

(define-private (foo-1 (impl-contract <trait-a>))
   (foo-2 impl-contract))

(define-private (foo-2 (impl-contract <trait-b>))
  true)
```

Should fail at analysis, but also the runtime type checks should be able to catch this. I think with these changes, that wouldn't be the case. Same goes for a drop of the trait entirely:

```clar
(define-private (foo-0 (impl-contract <trait-a>))
   (foo-1 impl-contract))

(define-private (foo-1 (impl-contract principal))
   (foo-2 impl-contract))

(define-private (foo-2 (impl-contract <trait-a>))
  true)
```

Reply:
Right, those will get caught by the analysis, so I didn't want to add complexity here to do the trait check again. You're saying we should do that check here as well?

Reply:
I think it's okay to leave this as "only checked during analysis", but we have to be very confident that the analysis pass is sufficiently protecting the runtime.

Reply:
I think its okay because that's exactly what was happening before, per the old comment:
> The code fetching and checking the trait is implemented in the contract_call eval function.
If an invalid call somehow snuck through, the worst case is that the runtime would hit an `UndefinedFunction` error when it tries to call the function.